File: typescript/list.ts
export class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }
}

export const buildList = (vals: number[]): ListNode | null => vals.length ? new ListNode(
    vals[0],
    buildList(vals.slice(1))
) : null;

export const compareList = (l1: ListNode | null, l2: ListNode | null): boolean => {
    if (!l1 && !l2) return true;
    if (!l1 || !l2) return false;
    return l1.val === l2.val && compareList(l1.next, l2.next);
};

File: python/list.py
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def buildList(vals: List[int]) -> Optional[ListNode]:
    match vals:
        case []:
            return None
        case [head, *tail]:
            head = ListNode(head)
            head.next = buildList(tail)
            return head


def compareList(l1: Optional[ListNode], l2: Optional[ListNode]) -> bool:
    while l1 and l2:
        if l1.val != l2.val:
            return False
        l1, l2 = l1.next, l2.next
    return l1 is None and l2 is None
